core/src/commonMain/kotlin/org/aquila3d/core/math/Quaternion.kt

```diff
@@ -0,0 +1,324 @@
+        val yNew = y / divisor
+        val zNew = z / divisor
+        return Vector3(xNew, yNew, zNew)
+    }
+
+    /**
+     * Gets the angle of rotation in radians for this quaternion.
+     *
+     * @return The angle of rotation in [Radians].
+     */
+    fun getAngle(): Radians {
+        return Radians(2.0 * acos(w))
+    }
+
+    /**
+     * Calculates the magnitude (norm) of this [Quaternion]
+     *
+     * @return The magnitude (norm).
+     */
+    fun magnitude(): Double {
+        return sqrt(w * w + x * x + y * y + z * z)
+    }
+
+    fun add(other: Quaternion): Quaternion {
+        w += other.w
+        x += other.x
+        y += other.y
+        z += other.z
+        return this
+    }
+
+    fun subtract(other: Quaternion): Quaternion {
+        w -= other.w
+        x -= other.x
+        y -= other.y
+        z -= other.z
+        return this
+    }
+
+    fun multiply(scale: Float): Quaternion {
+        return multiply(scale.toDouble())
+    }
+
+    fun multiply(scale: Double): Quaternion {
+        w *= scale
+        x *= scale
+        y *= scale
+        z *= scale
+        return this
+    }
+
+    fun multiply(vector3: Vector3): Quaternion {
+        val wNew = -(x * vector3.x + y * vector3.y + z * vector3.z)
+        val xNew = w * vector3.x + y * vector3.z - z * vector3.y
+        val yNew = w * vector3.y + z * vector3.x - x * vector3.z
+        val zNew = w * vector3.z + x * vector3.y - y * vector3.x
+        w = wNew
+        x = xNew
+        y = yNew
+        z = zNew
+        return this
+    }
+
+    fun multiply(other: Quaternion): Quaternion {
+        val wNew = w * other.w - x * other.x - y * other.y - z * other.z
+        val xNew = w * other.x + x * other.w + y * other.z - z * other.y
+        val yNew = w * other.y + y * other.w + z * other.x - x * other.z
+        val zNew = w * other.z + z * other.w + x * other.y - y * other.x
+        w = wNew
+        x = xNew
+        y = yNew
+        z = zNew
+        return this
+    }
+
+    fun divide(scale: Float): Quaternion {
+        return divide(scale.toDouble())
+    }
+
+    fun divide(scale: Double): Quaternion {
+        w /= scale
+        x /= scale
+        y /= scale
+        z /= scale
+        return this
+    }
+
+    fun normalize(): Quaternion {
+        val magnitude = magnitude()
+        return divide(magnitude)
+    }
+
+    fun invert(): Quaternion {
+        x = -x
+        y = -y
+        z = -z
+        return this
+    }
+
+    /**
+     * Rotates [vector3] by the rotation represented in this [Quaternion]. Mathematically this is equivalent to:
+     *
+     * **v'** = **q** * **v** * ~**q** where ~**q** is the conjugate or inverse of this [Quaternion].
+     */
+    fun rotate(vector3: Vector3): Vector3 {
+        val rotation = Quaternion(this)
+        val conj = Quaternion(this).invert()
+        return (rotation.multiply(vector3).multiply(conj).getAxis())
+    }
+
+    override fun toString(): String {
```

Comment:
You can remove the equality, hash, and toString in favor a kotlin data class which will provide this for you